Support __notes__ in pytest.raises

I would like `pytest.raises` to have some form of support for `__notes__`.

#### What's the problem this feature will solve?

I am currently looking into replacing dynamic patching of error messages with the new `__notes__` feature added in [PEP-678](https://peps.python.org/pep-0678/).

However, once I use `__notes__` my checks using `pytest.raises` now fail, and I'd like them to not. Or at least, I would like for pytest to have some features for matching these errors. Since this is a language level feature for python, I think it's usage should be covered by `pytest`.

#### Describe the solution you'd like

I would like the regex from `pytest.raises` to match against the notes as well as the body of the error. It would be fine for this to be specified with a keyword argument.

Example of usage:

```python
import pytest

def foo():
    assert False

def bar():
    try:
        foo()
    except AssertionError as e:
        e.add_note("Raised by bar")
        raise

with pytest.raises(AssertionError, match="Raised by bar"):
    bar()
```

<!-- Provide examples of real-world use cases that this would enable and how it solves the problem described above. -->

#### Alternative Solutions
<!-- Have you tried to workaround the problem using a pytest plugin or other tools? Or a different approach to solving this issue? Please elaborate here. -->

##### Do nothing

I could manually check for notes, but I don't think python makes it obvious which part of an exception is from a note. Here is a small example of what a note looks like:

<details>
<summary> setup </summary>

```python
# Python 3.11.4 | packaged by conda-forge | (main, Jun 10 2023, 18:10:28) [Clang 15.0.7 ] on darwin
# Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> 
>>> def add_note(err: BaseException, msg: str) -> None:
...     if sys.version_info < (3, 11):
...         err.__notes__ = getattr(err, "__notes__", []) + [msg]
...     else:
...         err.add_note(msg)
... 
>>> def foo():
...     raise AssertionError("foo is a bad function")
... 
>>> try:
...     foo()
... except AssertionError as e:
...     add_note(e, "yeah, foo is awful")
...     raise e
```
</details>

```pytb
Traceback (most recent call last):
  File "<stdin>", line 5, in <module>
  File "<stdin>", line 2, in <module>
  File "<stdin>", line 2, in foo
AssertionError: foo is a bad function
yeah, foo is awful
```

I think this would make it quite difficult for someone receiving an exception with notes from an upstream library to know how to match against it.

I'm currently handling this with a helper function:

```python
def check_error_or_notes_match(e: pytest.ExceptionInfo, pattern: str):
    """
    Checks whether the printed error message or the notes contains the given pattern.

    DOES NOT WORK IN IPYTHON - because of the way IPython handles exceptions
    """
    import traceback

    message = "".join(traceback.format_exception_only(e.type, e.value))
    assert re.search(
        pattern, message
    ), f"Could not find pattern: '{pattern}' in error:\n\n{message}\n"
```


#### Additional context
<!-- Add any other context, links, etc. about the feature here. -->

This line would have to change: https://github.com/pytest-dev/pytest/blob/b20e7f6d0cd85f439d7963b4ae7c3c3fc908452d/src/_pytest/_code/code.py#L707.

--- a/AUTHORS
+++ b/AUTHORS
@@ -168,6 +168,7 @@ Ian Bicking
 Ian Lesperance
 Ilya Konstantinov
 Ionuț Turturică
+Isaac Virshup
 Itxaso Aizpurua
 Iwan Briquemont
 Jaap Broekhuizen

--- a/changelog/11227.improvement.rst
+++ b/changelog/11227.improvement.rst
@@ -1,0 +1,1 @@
+Allow :func:`pytest.raises` ``match`` argument to match against `PEP-678 <https://peps.python.org/pep-0678/>` ``__notes__``.

--- a/src/_pytest/_code/code.py
+++ b/src/_pytest/_code/code.py
@@ -704,7 +704,12 @@ class ExceptionInfo(Generic[E]):
         If it matches `True` is returned, otherwise an `AssertionError` is raised.
         """
         __tracebackhide__ = True
-        value = str(self.value)
+        value = "\n".join(
+            [
+                str(self.value),
+                *getattr(self.value, "__notes__", []),
+            ]
+        )
         msg = f"Regex pattern did not match.\n Regex: {regexp!r}\n Input: {value!r}"
         if regexp == value:
             msg += "\n Did you mean to `re.escape()` the regex?"

--- a/src/_pytest/python_api.py
+++ b/src/_pytest/python_api.py
@@ -843,7 +843,8 @@ def raises(  # noqa: F811
         >>> with pytest.raises(ValueError, match=r'must be \d+$'):
         ...     raise ValueError("value must be 42")
 
-    The ``match`` argument searches the formatted exception string, which includes any ``__notes__``:
+    The ``match`` argument searches the formatted exception string, which includes any
+    `PEP-678 <https://peps.python.org/pep-0678/>` ``__notes__``:
 
         >>> with pytest.raises(ValueError, match=r'had a note added'):  # doctest: +SKIP
         ...    e = ValueError("value must be 42")

--- a/src/_pytest/python_api.py
+++ b/src/_pytest/python_api.py
@@ -843,6 +843,13 @@ def raises(  # noqa: F811
         >>> with pytest.raises(ValueError, match=r'must be \d+$'):
         ...     raise ValueError("value must be 42")
 
+    The ``match`` argument searches the formatted exception string, which includes any ``__notes__``:
+
+        >>> with pytest.raises(ValueError, match=r'had a note added'):  # doctest: +SKIP
+        ...    e = ValueError("value must be 42")
+        ...    e.add_note("had a note added")
+        ...    raise e
+
     The context manager produces an :class:`ExceptionInfo` object which can be used to inspect the
     details of the captured exception::
 

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -1,3 +1,5 @@
+from __future__ import annotations
+
 import importlib
 import io
 import operator
@@ -7,10 +9,7 @@ import sys
 import textwrap
 from pathlib import Path
 from typing import Any
-from typing import Dict
-from typing import Tuple
 from typing import TYPE_CHECKING
-from typing import Union
 
 import _pytest._code
 import pytest
@@ -802,7 +801,7 @@ raise ValueError()
         )
         excinfo = pytest.raises(ValueError, mod.entry)
 
-        styles: Tuple[_TracebackStyle, ...] = ("long", "short")
+        styles: tuple[_TracebackStyle, ...] = ("long", "short")
         for style in styles:
             p = FormattedExcinfo(style=style)
             reprtb = p.repr_traceback(excinfo)
@@ -929,7 +928,7 @@ raise ValueError()
         )
         excinfo = pytest.raises(ValueError, mod.entry)
 
-        styles: Tuple[_TracebackStyle, ...] = ("short", "long", "no")
+        styles: tuple[_TracebackStyle, ...] = ("short", "long", "no")
         for style in styles:
             for showlocals in (True, False):
                 repr = excinfo.getrepr(style=style, showlocals=showlocals)
@@ -1091,7 +1090,7 @@ raise ValueError()
             for funcargs in (True, False)
         ],
     )
-    def test_format_excinfo(self, reproptions: Dict[str, Any]) -> None:
+    def test_format_excinfo(self, reproptions: dict[str, Any]) -> None:
         def bar():
             assert False, "some error"
 
@@ -1399,7 +1398,7 @@ raise ValueError()
 @pytest.mark.parametrize("encoding", [None, "utf8", "utf16"])
 def test_repr_traceback_with_unicode(style, encoding):
     if encoding is None:
-        msg: Union[str, bytes] = "☹"
+        msg: str | bytes = "☹"
     else:
         msg = "☹".encode(encoding)
     try:

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -1648,3 +1648,38 @@ def test_hidden_entries_of_chained_exceptions_are_not_shown(pytester: Pytester) 
         ],
         consecutive=True,
     )
+
+
+@pytest.mark.skip("sys.version_info < (3,11)")
+@pytest.mark.parametrize(
+    "error,notes,match",
+    [
+        (Exception("test"), [], "test"),
+        (AssertionError("foo"), ["bar"], "bar"),
+        (AssertionError("foo"), ["bar", "baz"], "bar"),
+        (AssertionError("foo"), ["bar", "baz"], "baz"),
+    ],
+)
+def test_check_error_notes_success(error, notes, match):
+    for note in notes:
+        error.add_note(note)
+
+    with pytest.raises(Exception, match=match):
+        raise error
+
+
+@pytest.mark.skip("sys.version_info < (3,11)")
+@pytest.mark.parametrize(
+    "error, notes, match",
+    [
+        (Exception("test"), [], "foo"),
+        (AssertionError("foo"), ["bar"], "baz"),
+    ],
+)
+def test_check_error_notes_failure(error, notes, match):
+    for note in notes:
+        error.add_note(note)
+
+    with pytest.raises(AssertionError):
+        with pytest.raises(type(error), match=match):
+            raise error

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -1650,7 +1650,14 @@ def test_hidden_entries_of_chained_exceptions_are_not_shown(pytester: Pytester) 
     )
 
 
-@pytest.mark.skip("sys.version_info < (3,11)")
+def add_note(err: BaseException, msg: str) -> None:
+    """Adds a note to an exception inplace."""
+    if sys.version_info < (3, 11):
+        err.__notes__ = getattr(err, "__notes__", []) + [msg]  # type: ignore[attr-defined]
+    else:
+        err.add_note(msg)
+
+
 @pytest.mark.parametrize(
     "error,notes,match",
     [
@@ -1662,23 +1669,23 @@ def test_hidden_entries_of_chained_exceptions_are_not_shown(pytester: Pytester) 
 )
 def test_check_error_notes_success(error, notes, match):
     for note in notes:
-        error.add_note(note)
+        add_note(error, note)
 
     with pytest.raises(Exception, match=match):
         raise error
 
 
-@pytest.mark.skip("sys.version_info < (3,11)")
 @pytest.mark.parametrize(
     "error, notes, match",
     [
         (Exception("test"), [], "foo"),
         (AssertionError("foo"), ["bar"], "baz"),
+        (AssertionError("foo"), ["bar"], "foo\nbaz"),
     ],
 )
 def test_check_error_notes_failure(error, notes, match):
     for note in notes:
-        error.add_note(note)
+        add_note(error, note)
 
     with pytest.raises(AssertionError):
         with pytest.raises(type(error), match=match):

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -2,6 +2,7 @@ import importlib
 import io
 import operator
 import queue
+import re
 import sys
 import textwrap
 from pathlib import Path
@@ -1665,6 +1666,8 @@ def add_note(err: BaseException, msg: str) -> None:
         (AssertionError("foo"), ["bar"], "bar"),
         (AssertionError("foo"), ["bar", "baz"], "bar"),
         (AssertionError("foo"), ["bar", "baz"], "baz"),
+        (ValueError("foo"), ["bar", "baz"], re.compile(r"bar\nbaz", re.MULTILINE)),
+        (ValueError("foo"), ["bar", "baz"], re.compile(r"BAZ", re.IGNORECASE)),
     ],
 )
 def test_check_error_notes_success(

--- a/testing/code/test_excinfo.py
+++ b/testing/code/test_excinfo.py
@@ -1667,7 +1667,7 @@ def add_note(err: BaseException, msg: str) -> None:
         (AssertionError("foo"), ["bar", "baz"], "baz"),
     ],
 )
-def test_check_error_notes_success(error, notes, match):
+def test_check_error_notes_success(error: Exception, notes: list[str], match: str) -> None:
     for note in notes:
         add_note(error, note)
 
@@ -1683,7 +1683,7 @@ def test_check_error_notes_success(error, notes, match):
         (AssertionError("foo"), ["bar"], "foo\nbaz"),
     ],
 )
-def test_check_error_notes_failure(error, notes, match):
+def test_check_error_notes_failure(error: Exception, notes: list[str], match: str) -> None:
     for note in notes:
         add_note(error, note)
 

